activerecord/lib/active_record/connection_adapters/abstract/quoting.rb

```diff
@@ -159,7 +146,8 @@ def quote_default_expression(value, column) # :nodoc:
         if value.is_a?(Proc)
           value.call
         else
-          value = lookup_cast_type(column.sql_type).serialize(value)
+          cast_type = column.respond_to?(:cast_type) ? column.cast_type : lookup_cast_type(column.sql_type)
```

Comment:
Why do we need to check the existence of `cast_type`. Is it to support different adapters?

Reply:
Sadly this method ends up receiving two different types of objects: both `Column` and `ColumnDefinition` (from schema migration code).

Reply:
I tried to go down that path earlier, but it got pretty messy, because:

1. That object doesn't have access to a connection.
2. When the object is created we don't appear to have the right type information (e.g., `type` might be `:primary_key`).

Given it only happens when modifying the schema, it seemed not worth making an invasive change there, but I can try to add something like that back for you to see if you want.

Reply:
I think we can solve that by adding another method in both Column and ColumnDefinition

```
class Column
  def cast_type_for(connection)
    cast_type
  end
end

class ColumnDefinition
  def cast_type_for(connection)
    connection.lookup_cast_type(sql_type)
  end
end
```

And just call that method here:

```suggestion
          cast_type = column.cast_type_for(self)
```

Reply:
But the point of this PR is to avoiding needing the `connection` at all in `Column`, so that modeling would defeat the purpose.

Side note: the schema definition code is pretty confusing here (and I think likely mistakenly so), because it seems to use `ColumnDefinition` in some places and `Column` in other places.

Reply:
Well, I guess we could still use `cast_type` directly in some places, but I think it makes it a lot easier to make this mistake again in the future. Note we already used to have `type_for_column` accept a `connection`, but that meant we needed to checkout a connection unnecessarily in some cases.

Reply:
In this case I don't think it is a problem because the only caller for that method is already the connection, so you already have one. All the other places can just use `cast_type`

Reply:
Other option is to create a Column based of a ColumnDefinition when visiting the schema. We have the connection that that point. This way `Column` really don't need to know anything about the connection. And `quote_default_expression` can always work with `Column` objects.

I'm trying to avoid `quote_default_expression` to keep working with mixed interfaces.

Reply:
But I can do that later. Let's move on with this PR now that I understand the reason for the change